### PR TITLE
legacypoliteia: Add importProposal function

### PR DESCRIPTION
**Pull request by thi4go**
This diff adds the implementation of the `importProposal` function.

### CODE REVIEW

File: politeiad/cmd/legacypoliteia/blobs.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+	"github.com/decred/politeia/politeiad/plugins/comments"
+	"github.com/decred/politeia/politeiad/plugins/ticketvote"
+	v1 "github.com/decred/politeia/politeiawww/api/ticketvote/v1"
+	"github.com/decred/politeia/politeiawww/client"
+)
+
+// serverPubkey is the former politeia public key from when it ran the git
+// backend. It is used to verify the auth details and cast vote details
+// signatures.
+const serverPubkey = "a70134196c3cdf3f85f8af6abaa38c15feb7bccf5e6d3db6212358363465e502"
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/blobs.go]: This is already defined in the `gitbe` package.

File: politeiad/cmd/legacypoliteia/blobs.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+// Ticketvote blobs
+
+// blobSaveCastVoteDetails saves the cast vote details blob to tstore for the
+// specified proposal token. It also verifies the signature of the cast vote
+// details struct.
+func blobSaveCastVoteDetails(cdv ticketvote.CastVoteDetails, newToken []byte, cmd *importCmd) error {
+	// Verify cast vote details signature.
+	err := client.CastVoteDetailsVerify(convertCastVoteDetailsToV1(cdv),
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/blobs.go]: You don't need to do any verification in this command.  This is all handled by the convert command.

File: politeiad/cmd/legacypoliteia/blobs.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	"github.com/decred/politeia/politeiawww/client"
+)
+
+// serverPubkey is the former politeia public key from when it ran the git
+// backend. It is used to verify the auth details and cast vote details
+// signatures.
+const serverPubkey = "a70134196c3cdf3f85f8af6abaa38c15feb7bccf5e6d3db6212358363465e502"
+
+// Ticketvote blobs
+
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/blobs.go]: This file can be simplified into a single function.

`saveBlob(data interface{}, dataDescriptor string) error`

File: politeiad/cmd/legacypoliteia/blobs.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+	}
+
+	data, err := json.Marshal(cdv)
+	if err != nil {
+		return err
+	}
+	hint, err := json.Marshal(
+		store.DataDescriptor{
+			Type:       store.DataTypeStructure,
+			Descriptor: ticketvote.PluginID + "-castvote-v1",
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/blobs.go]: Don't define inline.

File: politeiad/cmd/legacypoliteia/blobs.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+		})
+	if err != nil {
+		return err
+	}
+	be := store.NewBlobEntry(hint, data)
+
+	err = cmd.tstore.BlobSave(newToken, be)
+	if err != nil && err.Error() == "duplicate payload" {
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/blobs.go]: Don't define inline.  We don't hardcode errors like this either.

File: politeiad/cmd/legacypoliteia/proposal.go [hunk 6]
```diff
@@ -28,9 +28,8 @@ type proposal struct {
 	// usermd plugin
 	StatusChanges []usermd.StatusChangeMetadata
 
-	// ticketvote plugin. These fields may be nil depending on the proposal,
-	// i.e. abandoned proposals will not have ticketvote data.
-	VoteMetadata *ticketvote.VoteMetadata
+	// ticketvote plugin
+	VoteMetadata ticketvote.VoteMetadata
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/proposal.go]: Don't change this.  A VoteMetadata is not going to exist for every proposal.

File: politeiad/cmd/legacypoliteia/importcmd.go [hunk 7]
```diff
@@ -134,24 +141,240 @@ func importProposals(legacyDir string, cmd *importCmd) error {
...
-// The parent token of all RFP submissions are updated with the tstore RFP
-// parent token.
-func importProposal(legacyDir, gitToken string, cmd *importCmd) {}
+// 4. Import comment blobs and vote blobs concurrently. This is the most costly
+//    process of importing, and is why it is done concurrently.
+//
+// 5. Add git token to tstore token mapping to the memory cache for all RFP
+// 	  parent proposals. The parent token of all RFP submissions are updated
+//    with the tstore RFP parent token.
+func importProposal(legacyDir, gitToken string, cmd *importCmd) error {
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/importcmd.go]: This function needs to be easier to read and review.  I don't want to have to go searching into nested functions to find when something is being saved.

https://github.com/decred/politeia/blob/fa7b7ad91df5b1ee23a26e1358d950dd222e4dbf/politeiad/cmd/legacypoliteia/convertcmd.go#L94-L176

In the above code, each function corresponds to a field on the proposal and they are called in the same order that the fields are listed in the proposal.  I can easily go down and audit this code.  If I want to see exactly how one of the fields is being derived, I can look into the convert function for further details.

**Review by lukebp — CHANGES_REQUESTED** [review]

File: politeiad/cmd/legacypoliteia/importcmd.go (hunk 7 above)
Comment (lukebp) [reply politeiad/cmd/legacypoliteia/importcmd.go]: It  would also help with readability if you used the naming convention `import`, to be consistent with the command name, and put all of the import functions into a `import.go` file.

Ex. `importRecordMetadata()`, `importFiles()`, etc

File: politeiad/cmd/legacypoliteia/proposal.go (hunk 6 above)
Comment (thi4go) [reply politeiad/cmd/legacypoliteia/proposal.go]: oh my bad, rebase issues

File: politeiad/cmd/legacypoliteia/import.go [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+// metadatas before saving, if needed.
+func importRecord(p *proposal, tstoreToken []byte, cmd *importCmd) error {
+	// Prepare vote metadata file if needed.
+	if p.VoteMetadata.LinkBy != 0 || p.VoteMetadata.LinkTo != "" {
+		b, err := json.Marshal(p.VoteMetadata)
+		if err != nil {
+			return err
+		}
+		p.Files = append(p.Files, backend.File{
+			Name:    "votemetadata.json",
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/import.go]: Use the plugin variable.

File: politeiad/cmd/legacypoliteia/import.go [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+		err := saveBlob(vote, dataDescriptorVoteDetails, tstoreToken, cmd)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func importJournals(prop *proposal, tstoreToken []byte, cmd *importCmd) error {
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/import.go]: These need to be broken up into importCastVotes, importCommentAdds, importCommentDels, etc.

File: politeiad/cmd/legacypoliteia/import.go [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+		err := importVotes(p, token, cmd)
+		if err != nil {
+			e := fmt.Errorf("saveVotes thread err: %w", err)
+			panic(e)
+		}
+	}(prop, tstoreToken, cmd)
+
+	// Spin thread for saving comment blobs.
+	wg.Add(1)
+	go func(p *proposal, token []byte, cmd *importCmd) {
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/import.go]: I would have a single saveBlobsConcurrently method that is generic and is used anytime you're saving blobs concurrently.

File: politeiad/cmd/legacypoliteia/proposal.go [hunk 11]
```diff
@@ -28,8 +28,7 @@ type proposal struct {
 	// usermd plugin
 	StatusChanges []usermd.StatusChangeMetadata
 
-	// ticketvote plugin. These fields may be nil depending on the proposal,
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/proposal.go]: Put comment back in.

File: politeiad/cmd/legacypoliteia/import.go [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+	if err != nil {
+		return err
+	}
+
+	// Replace gitbe tokens with the new tstorebe token.
+	err = updateProposalWithTstoreToken(prop, hex.EncodeToString(token), cmd)
+	if err != nil {
+		return err
+	}
+
```
Comment (lukebp) [inline politeiad/cmd/legacypoliteia/import.go:93]: Any updates to the proposal should be done here.  I don't want the state changes nested inside of functions.  Doing it all in the same place makes it easier to audit. This function should read as:

1. Load the proposal
2. Create the tstore record
3. Update the proposal with the tstore token
4. Import the different parts of the proposal into tstore

**Review by lukebp — CHANGES_REQUESTED** [review]

File: politeiad/cmd/legacypoliteia/import.go (hunk 12 above)
Comment (lukebp) [reply politeiad/cmd/legacypoliteia/import.go:93]: If this is not possible for all scenerios then we should talk it through.

**Pull request closed without merging**